### PR TITLE
chore: extract --project-grep option for matching with regex

### PR DESCRIPTION
**Pull request by yury-s**
Reference https://github.com/microsoft/playwright/issues/15128

### CODE REVIEW

File: packages/playwright/src/program.ts [hunk 1]
```diff
@@ -65,7 +65,8 @@ function addListFilesCommand(program: Command) {
   const command = program.command('list-files [file-filter...]', { hidden: true });
   command.description('List files with Playwright Test tests');
   command.option('-c, --config <file>', `Configuration file, or a test directory with optional "playwright.config.{m,c}?{js,ts}"`);
-  command.option('--project <project-name...>', `Only run tests from the projects matching this regular expression (default: list all projects)`);
+  command.option('--project <project-name...>', `Only run tests from the specified list of projects (default: list all projects)`);
+  command.option('--project-grep <project-name...>', `Only run tests from the projects matching this regular expression (default: list all projects)`);
```
Comment (dgozman) [inline packages/playwright/src/program.ts:69]: Make this one a single-value only?

File: packages/playwright/src/program.ts [hunk 2]
```diff
@@ -328,7 +330,8 @@ const testOptions: [string, string][] = [
   ['--no-deps', 'Do not run project dependencies'],
   ['--output <dir>', `Folder for output artifacts (default: "test-results")`],
   ['--pass-with-no-tests', `Makes test run succeed even if no tests were found`],
-  ['--project <project-name...>', `Only run tests from the projects matching this regular expression (default: run all projects)`],
+  ['--project <project-name...>', `Only run tests from the specified list of projects(default: run all projects)`],
+  ['--project-grep <pattern...>', `Only run tests from the projects matching this regular expression (default: run all projects)`],
```
Comment (dgozman) [inline packages/playwright/src/program.ts:334]: Same here, single value only?

File: packages/playwright/src/util.ts [hunk 3]
```diff
@@ -91,6 +91,12 @@ export function createFileFiltersFromArguments(args: string[]): TestFileFilter[]
   });
 }
 
+export function createProjectMatcher(projectNames?: string[], projectGrep?: string[]): Matcher {
+  const names = projectNames || [];
+  const regexps = (projectGrep || []).map(grep => forceRegExp(grep));
+  return (project: string) => (!names.length && !regexps.length) || names.includes(project) || regexps.some(re => re.test(project));
```
Comment (dgozman) [inline packages/playwright/src/util.ts:97]: I'd think that both project name filter and project grep filter should match, not just one of them. Or maybe disallow that case in the CLI right away?

File: tests/playwright-test/config.spec.ts [hunk 4]
```diff
@@ -259,33 +260,31 @@ test('should filter by project and parse as RegExp', async ({ runInlineTest }) =
...
-  expect(new Set(result.outputLines)).toEqual(new Set(['foobar']));
+    'a.test.ts': `
+      import { test, expect } from '@playwright/test';
+      test('pass', async ({}, testInfo) => {
+        console.log(testInfo.project.name);
+      });
+    `
+  }, { '--project-grep': ['aaa', 'bbb', 'SUite1'] });
+  expect(exitCode).toBe(1);
+  expect(output).toContain('Project(s) matching "aaa", "bbb" not found. Available named projects: "suite1", "suite2"');
```
Comment (dgozman) [inline tests/playwright-test/config.spec.ts:287]: Why is this an error? Perhaps only show an error when no projects matched at all?

File: tests/playwright-test/config.spec.ts [hunk 5]
```diff
@@ -304,7 +303,7 @@ test('should filter by project and allow passing RegExp start/end flags', async
       test('one', async ({}) => {
         console.log('%%' + test.info().project.name);
       });    `
-  }, { project: '/fooBar$/' });
+  }, { '--project-grep': '/fooBar$/' });
```
Comment (dgozman) [inline tests/playwright-test/config.spec.ts:306]: I'd say with a separate option we can probably omit `//`.

File: packages/playwright/src/runner/projectUtils.ts [hunk 6]
```diff
@@ -19,38 +19,56 @@ import path from 'path';
...
     if (!names.length)
       throw new Error(`No named projects are specified in the configuration file`);
-    const unknownProjectNames = Array.from(unmatchedProjectFilters.values()).map(n => `"${n}"`).join(', ');
-    throw new Error(`Project(s) ${unknownProjectNames} not found. Available named projects: ${names.map(name => `"${name}"`).join(', ')}`);
+    if (unmatchedProjectNames.size) {
+      const unknownProjectNames = Array.from(unmatchedProjectNames.values()).map(n => `"${n}"`).join(', ');
+      throw new Error(`Project(s) ${unknownProjectNames} not found. Available named projects: ${names.map(name => `"${name}"`).join(', ')}`);
+    }
+    const unknownProjectGreps = Array.from(unmatchedProjectGreps.values()).map(n => `"${n}"`).join(', ');
+    throw new Error(`Project(s) matching ${unknownProjectGreps} not found. Available named projects: ${names.map(name => `"${name}"`).join(', ')}`);
```
Comment (dgozman) [inline packages/playwright/src/runner/projectUtils.ts:71]: I'd drop this logic, and only throw when `result.length === 0`. WDYT?

File: tests/playwright-test/config.spec.ts (hunk 5 above)
Comment (yury-s) [reply tests/playwright-test/config.spec.ts:306]: this matches our --grep behavior

File: tests/playwright-test/config.spec.ts (hunk 4 above)
Comment (yury-s) [reply tests/playwright-test/config.spec.ts:287]: The idea is to do the same as with --project, i.e. each specified regex should match something. But with single regex it should not be a problem.

File: packages/playwright/src/program.ts [hunk 7]
```diff
@@ -152,6 +157,9 @@ async function runTests(args: string[], opts: { [key: string]: any }) {
 }
 
 async function listTestFiles(opts: { [key: string]: any }) {
+  if (opts.project && opts.projectGrep)
```
Comment (dgozman) [inline packages/playwright/src/program.ts:160]: I guess this method should also respect `opts.projectGrep` when listing?

File: packages/playwright/src/runner/projectUtils.ts [hunk 8]
```diff
@@ -19,39 +19,49 @@ import path from 'path';
...
     return [...projects];
-  const unmatchedProjectFilters = new Set<string>(projectNames);
+
+  const names = projects.map(p => p.project.name).filter(name => !!name);
+  if (!names.length)
+    throw new Error(`No named projects are specified in the configuration file`);
+
+  if (projectGrep) {
+    const regex = forceRegExp(projectGrep);
+    const result = projects.filter(project => regex.test(project.project.name));
```
Comment (dgozman) [inline packages/playwright/src/runner/projectUtils.ts:37]: When testing on the same RegExp instance, you need to reset `lastIndex` property.

File: packages/playwright/src/util.ts [hunk 9]
```diff
@@ -91,6 +91,12 @@ export function createFileFiltersFromArguments(args: string[]): TestFileFilter[]
   });
 }
 
+export function createProjectMatcher(projectNames?: string[], projectGrep?: string): Matcher {
+  const names = projectNames || [];
+  const regexp = projectGrep ? forceRegExp(projectGrep) : undefined;
+  return (project: string) => (!!names.length && names.includes(project)) || (!!regexp && regexp.test(project));
```
Comment (dgozman) [inline packages/playwright/src/util.ts:97]: The other place uses `toLowerCase` on the project names filter. I also wonder why do we have two separate filters in our codebase?

**Review by dgozman — APPROVED** [review]

File: packages/playwright/src/util.ts (hunk 9 above)
Comment (yury-s) [reply packages/playwright/src/util.ts:97]: Good question. Change this place to use filterProjects too.

**Pull request merged**